# Sagar patil
# Bhushan Savarkar
# Prarthana Saikia
# Monalisa das
# Soham Chowdhari

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
import statsmodels.formula.api as sm

In [3]:
data = pd.read_csv("E:\\Praxis\\Customer Analytics\\Blend test\\FINALdatablindtest.csv")
data.head()

Resp No  PANEL  CENTRE  MAIN_BRND  AGE  AGES  Plcmnt_Ordr  Prod  Q2A  Q2B  \
0        1      2       1          2   34     2            1     2    7    3   
1        1      2       1          2   34     2            2     3    8    5   
2        1      2       1          2   34     2            3     1    4    5   
3        2      3       1          2   26     1            1     3    7    4   
4        2      3       1          2   26     1            2     1    9    2   

          ...          Q5A_att5  Q5A_att6  Q5A_att7  Q5A_att8  Q5B_att1  \
0         ...                 7         7         7         7         3   
1         ...                 8         7         8         8         5   
2         ...                 7         7         6         7         5   
3         ...                 7         8        10         8         3   
4         ...                 9        10        10         9         3   

   Q5B_att2  Q5B_att3  Q5B_att4  Q5B_att5  Q6_Int_p (Y=1,N=2)  
0         4         4         4         3                   2  
1         5         5         5         3                   2  
2         5         3         5         5                   2  
3         3         3         3         3                   1  
4         3         4         4         3                   1  

[5 rows x 26 columns]

In [4]:
result = sm.ols(formula="Q5A_att1~Q5A_att2",data=data).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:               Q5A_att1   R-squared:                       0.523
Model:                            OLS   Adj. R-squared:                  0.523
Method:                 Least Squares   F-statistic:                     2496.
Date:                Wed, 13 Jun 2018   Prob (F-statistic):               0.00
Time:                        21:25:51   Log-Likelihood:                -3593.1
No. Observations:                2280   AIC:                             7190.
Df Residuals:                    2278   BIC:                             7202.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.6425      0.102     25.888      0.0

In [5]:
result = sm.ols(formula="Q5A_att1~Q5A_att2+Q5A_att3+Q5A_att4+Q5A_att5+Q5A_att6+Q5A_att7+Q5A_att8",data=data).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:               Q5A_att1   R-squared:                       0.775
Model:                            OLS   Adj. R-squared:                  0.774
Method:                 Least Squares   F-statistic:                     1117.
Date:                Wed, 13 Jun 2018   Prob (F-statistic):               0.00
Time:                        21:25:54   Log-Likelihood:                -2736.5
No. Observations:                2280   AIC:                             5489.
Df Residuals:                    2272   BIC:                             5535.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5755      0.082      7.042      0.0

Q5A_att7 which is after-taste is not significant as its p-value is > 0.05

In [6]:
result_2 = sm.ols(formula="Q5A_att1~Q5A_att2+Q5A_att3+Q5A_att4+Q5A_att5+Q5A_att6+Q5A_att8",data=data).fit()
print(result_2.summary())

                            OLS Regression Results                            
Dep. Variable:               Q5A_att1   R-squared:                       0.775
Model:                            OLS   Adj. R-squared:                  0.774
Method:                 Least Squares   F-statistic:                     1302.
Date:                Wed, 13 Jun 2018   Prob (F-statistic):               0.00
Time:                        21:25:56   Log-Likelihood:                -2738.0
No. Observations:                2280   AIC:                             5490.
Df Residuals:                    2273   BIC:                             5530.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5892      0.081      7.241      0.0

Now all variables are significant as each one is having a p-value < 0.05.




R2 and Adj. R2 have not reduced significantly.


Overall model is significant as p-value of F statistic is almost equal to zero.

Equation to predict overall likebility:

overall likebility = 0.5892 + 0.1604(Aroma) + 0.2880(Taste) + 0.1102(Smoothness) + 0.0835(Flavour) + 0.0937(Throat_feel) + 0.2005(Mouth-feel)

In [7]:
from sklearn.decomposition import FactorAnalysis, PCA
from scipy.linalg import eig
from sklearn import preprocessing

 Extracting required predictors in separate dataframe

In [8]:
data_new = data.loc[:,'Q5A_att2':'Q5A_att8']

In [9]:
data_new.corr()

Q5A_att2  Q5A_att3  Q5A_att4  Q5A_att5  Q5A_att6  Q5A_att7  Q5A_att8
Q5A_att2  1.000000  0.699165  0.650235  0.699211  0.644325  0.658032  0.652336
Q5A_att3  0.699165  1.000000  0.777731  0.783251  0.749416  0.790696  0.796745
Q5A_att4  0.650235  0.777731  1.000000  0.730227  0.783715  0.758622  0.761744
Q5A_att5  0.699211  0.783251  0.730227  1.000000  0.742777  0.772320  0.768208
Q5A_att6  0.644325  0.749416  0.783715  0.742777  1.000000  0.795996  0.814465
Q5A_att7  0.658032  0.790696  0.758622  0.772320  0.795996  1.000000  0.830628
Q5A_att8  0.652336  0.796745  0.761744  0.768208  0.814465  0.830628  1.000000

In [10]:
from scipy.stats import bartlett as bar

Standardizing the data

In [11]:
data_std = pd.DataFrame(preprocessing.scale(data_new))

# PCA

In [18]:
pca = PCA(n_components=3)

3 principal components are extracted which are explaining approximately 88% of variance

In [19]:
pc = pd.DataFrame(pca.fit_transform(data_std))

In [20]:
pca.explained_variance_ratio_

array([ 0.78316344,  0.06156255,  0.03938998])

In [21]:
pca.explained_variance_

array([ 5.48454962,  0.43112695,  0.27585087])

These are the Eigen values.

In [22]:
components = pd.DataFrame(pca.components_).T
components

0         1         2
0 -0.342547 -0.865037 -0.196473
1 -0.386484 -0.036997  0.128842
2 -0.377017  0.172262 -0.687332
3 -0.379144 -0.137822  0.536239
4 -0.382102  0.294623 -0.276010
5 -0.387534  0.217187  0.265839
6 -0.388863  0.260194  0.194854

In [25]:
cov_mat = data_std.cov()

In [26]:
eig_vector = eig(cov_mat)[1]
eig_value = eig(cov_mat)[0]
eig_vector

array([[ 0.34254692,  0.86503651,  0.1964729 , -0.29343641,  0.07394761,
         0.04988682, -0.04135178],
       [ 0.3864845 ,  0.036997  , -0.12884179,  0.48868823,  0.53533552,
        -0.35198195,  0.42821624],
       [ 0.37701679, -0.17226206,  0.68733232,  0.44690016, -0.10998941,
         0.23825389, -0.29525593],
       [ 0.37914438,  0.13782232, -0.53623914,  0.34068545, -0.65438562,
         0.06650644, -0.03149456],
       [ 0.38210157, -0.29462349,  0.2760099 , -0.45995263, -0.36931219,
        -0.43338598,  0.39400837],
       [ 0.38753427, -0.21718663, -0.26583937, -0.24740552,  0.27192487,
        -0.29736679, -0.7130196 ],
       [ 0.38886347, -0.26019379, -0.19485387, -0.29415411,  0.23936331,
         0.7322442 ,  0.25122373]])

In [27]:
eig_value

array([ 5.48454962+0.j,  0.43112695+0.j,  0.27585087+0.j,  0.25232410+0.j,
        0.22033185+0.j,  0.15978157+0.j,  0.17910657+0.j])

# Factor Analysis with 3 factor solution

In [31]:
%load_ext rpy2.ipython

In [32]:
import rpy2
import warnings
warnings.filterwarnings('ignore')

In [33]:
%R -i data_std

In [34]:
%R require(psych)
%R pca1 = principal(data_std,nfactors=3,scores=TRUE)

R object with classes: ('psych', 'principal') mapped to:
<ListVector - Python:0x000002160310D248 / R:0x0000021603E073D0>
[FloatVector, StrVector, IntVector, FloatVector, ..., Matrix, Matrix, Matrix, Matrix]
  values: <class 'rpy2.robjects.vectors.FloatVector'>
  R object with classes: ('numeric',) mapped to:
<FloatVector - Python:0x000002160129DA88 / R:0x0000021603291C10>
[5.482144, 0.430938, 0.275730, 0.252213, 0.220235, 0.179028, 0.159711]
  rotation: <class 'rpy2.robjects.vectors.StrVector'>
  R object with classes: ('character',) mapped to:
<StrVector - Python:0x0000021603119748 / R:0x0000021603B2E3E0>
['varimax']
  n.obs: <class 'rpy2.robjects.vectors.IntVector'>
  R object with classes: ('integer',) mapped to:
<IntVector - Python:0x0000021603119EC8 / R:0x0000021603D38B98>
[2,280]
  communality: <class 'rpy2.robjects.vectors.FloatVector'>
  R object with classes: ('numeric',) mapped to:
<FloatVector - Python:0x0000021603119C08 / R:0x0000021603285138>
[0.976375, 0.824036, 0.922291, 0.875533, 0.858814, 0.863137, 0.868625]
...
  fit: <class 'rpy2.robjects.vectors.Matrix'>
  R object with classes: ('matrix',) mapped to:
<Matrix - Python:0x0000021603119448 / R:0x0000021602A37898>
[0.665929, -0.234049, -0.708350, 0.458581, ..., 0.141227, 0.588423, -0.418883, 0.691589]
  fit.off: <class 'rpy2.robjects.vectors.Matrix'>
  R object with classes: ('matrix',) mapped to:
<Matrix - Python:0x00000216030A5208 / R:0x00000216029E7CD0>
[2.593077, 0.370440, 0.370440, 0.418994, ..., 0.212871, 0.884116, 0.240772, 1.000000]
  fn: <class 'rpy2.robjects.vectors.Matrix'>
  R object with classes: ('loadings',) mapped to:
<Matrix - Python:0x0000021603119A48 / R:0x00000216025B97F0>
[0.328115, 0.644848, 0.358657, 0.769445, ..., 0.446708, 0.261051, 0.271301, 0.253222]
  Call: <class 'rpy2.robjects.vectors.Matrix'>
  R object with classes: ('matrix',) mapped to:
<Matrix - Python:0x0000021603119408 / R:0x0000021603F54068>
[0.042954, 0.273118, -0.685239, 0.375637, ..., -1.150503, 0.275628, 0.291286, -1.150503]

In [35]:
%R pca1$values

array([ 5.48214411,  0.43093786,  0.27572989,  0.25221344,  0.22023521,
        0.17902801,  0.15971149])

In [36]:
fa = %R pca1$Structure
fa = pd.DataFrame(fa,columns=['F1','F2','F3'], index=data_new.columns)
fa

F1        F2        F3
Q5A_att2  0.328115  0.305421  0.880587
Q5A_att3  0.644848  0.475510  0.426730
Q5A_att4  0.358657  0.816404  0.356568
Q5A_att5  0.769445  0.289724  0.446708
Q5A_att6  0.538378  0.707683  0.261051
Q5A_att7  0.736495  0.497100  0.271301
Q5A_att8  0.718772  0.536536  0.253222

- "fa" is the Rotated Component Matrix
- Every attribute is strongly correlated with one or other factor as follows:
- Q5A_att2-F3
- Q5A_att3-F1
- Q5A_att4-F2
- Q5A_att5-F1
- Q5A_att6-F2
- Q5A_att7-F1
- Q5A_att8-F1

- so we can interpret these as factors as:
- factor 1- combination of taste,flavour,after-taste,mouth-feel
- factor 2- combination of smootheness,throat-feel
- factor 3- aroma

- Now, we will use these factors to do regression.

In [37]:
PCs = %R pca1$scores

In [38]:
PCs = pd.DataFrame(PCs, columns=['PC1','PC2','PC3'])
PCs.loc[:,"Q5A_att1"] = data['Q5A_att1']
PCs.head()

PC1       PC2       PC3  Q5A_att1
0  0.042954 -0.052273 -0.821272         7
1  0.273118 -0.169071  0.346631         8
2 -0.685239  0.623113 -0.642846         7
3  0.375637  0.441877 -0.175951        10
4  0.812690  0.824105  0.257039         9

In [39]:
m = sm.ols(formula="Q5A_att1~PC1+PC2+PC3", data=PCs)
M = m.fit()
print(M.summary())

                            OLS Regression Results                            
Dep. Variable:               Q5A_att1   R-squared:                       0.767
Model:                            OLS   Adj. R-squared:                  0.766
Method:                 Least Squares   F-statistic:                     2493.
Date:                Wed, 13 Jun 2018   Prob (F-statistic):               0.00
Time:                        21:37:35   Log-Likelihood:                -2777.3
No. Observations:                2280   AIC:                             5563.
Df Residuals:                    2276   BIC:                             5586.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.5930      0.017    442.810      0.0